post-scan-actions/azure-python-promote-or-quarantine/promoteOrQuarantineFunction/ScanResultHandler/handler.py

```diff
@@ -0,0 +1,217 @@
+    promote_storage_connection_string = os.environ.get('PROMOTE_STORAGE_CONNECTION_STRING')
+
+    promote_mode = get_promote_mode()
+    quarantine_mode = get_quarantine_mode()
+
+    message = json.loads(message_body)
+    file_url = message['file_url']
+    (_, blob_container, blob_name) = parse_blob_information(file_url)
+
+    source_blob_service_client = BlobServiceClient(
+        account_url=get_blob_account_url(file_url),
+        credential=credential
+    )
+    blob_url_sas = get_blob_url_sas(source_blob_service_client, file_url)
+
+    protecting_blob_client = source_blob_service_client.get_blob_client(blob_container, blob_name)
+    existing_metadata = get_existing_metadata(protecting_blob_client)
+    existing_tags = get_existing_tags(protecting_blob_client)
+
+    result = message['scanning_result']
+    findings = result['Findings']
+    logging.info(f'findings: {json.dumps(findings)}')
+
+    operation = 'quarantine' if findings else 'promotion'
+    mode = quarantine_mode if findings else promote_mode
+    dest_storage_connection_string = quarantine_storage_connection_string if findings else promote_storage_connection_string
+    scan_result = 'malicious' if findings else 'no issues found'
+    scan_date = time.strftime('%Y/%m/%d %H:%M:%S', time.localtime(message['timestamp']))
+
+    if not dest_storage_connection_string:
+        print(f'Skip: No storage connection string specified for {operation}')
+        return
+
+    codes = result['Codes']
+    code = CODE_EMPTY
+    if len(codes) > 0:
+        code = CODE_SKIP_MULTIPLE if len(codes) > 1 else codes[0]
+    fssTags = {
+        'scanned': 'true',
+        'scanDate': scan_date,
+        'scanResult': scan_result,
+        'scanDetailCode': str(code),
+        'scanDetailMessage': CODE_MESSAGES.get(code, CODE_MESSAGES[CODE_MISC])
+    }
+    logging.info(f'FSS tags: {fssTags}')
+    metadata = compose_metadata(existing_metadata, fssTags)
+    tags = compose_tags(existing_tags, fssTags)
+
+    dest_blob_service_client = BlobServiceClient.from_connection_string(dest_storage_connection_string)
+    copy_object(
+        source_blob_url=blob_url_sas,
+        container=blob_container,
+        blob_name=blob_name,
+        metadata=metadata,
+        tags=tags,
+        dest_blob_service_client=dest_blob_service_client,
+    )
+
+    if mode == 'move':
+        protecting_blob_client.delete_blob()
+
+    logging.info(f'File {operation} successful (mode: {mode})')
+
+
+def copy_object(source_blob_url, container, blob_name, metadata, tags, dest_blob_service_client: BlobServiceClient):
+    container_client = dest_blob_service_client.get_container_client(container)
+    if not container_client.exists():
+        container_client.create_container()
+
+    blob_client = container_client.get_blob_client(blob_name)
+    copy_status = blob_client.start_copy_from_url(
+        source_url=source_blob_url,
+        metadata=metadata,
+        tags=tags,
+        requires_sync=True

```

Comment:
If the function copies large size blob in sync mode would it encounter function timeout?

Reply:
It's quite tricky. If we don't block the copy here, we'll don't know when we can delete the file from protecting storage account. The solution would be using async event loop pattern. AWS's plugin has the same issue, since we didn't receive requests for that, we can leave it as enhancement.
The function now has 30 seconds timeout, and the consumed messages are locked for 1 minute in the scan result topic.